### PR TITLE
More dtype tracking

### LINKED ISSUE #89: dtype promotion rules are not quite right for element-wise ufuncs
Many ufuncs actually always converge this argument to float, e.g., consider:
```
In [13]: values = np.random.randint(10, size=(4, 6))

In [14]: data = da.from_array(values, blockshape=(2, 2))

In [15]: data.dtype
Out[15]: dtype('int64')

In [16]: da.sin(data).dtype
Out[16]: dtype('int64')

In [17]: da.sin(data).compute()
Out[17]:
array([[ 0.14112001,  0.14112001,  0.14112001,  0.90929743,  0.41211849,
         0.        ],
       [ 0.6569866 ,  0.14112001, -0.2794155 , -0.2794155 ,  0.14112001,
         0.90929743],
       [ 0.        ,  0.90929743,  0.84147098, -0.7568025 ,  0.98935825,
         0.90929743],
       [-0.95892427,  0.90929743,  0.14112001,  0.14112001,  0.        ,
         0.        ]])
```

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #89

As @shoyer brings up, #87 is incomplete and contains wrong results.  We'll fix some here.  Presumably we'll fail to find a few.  Is the current approach best or should we always rely on some small amount of computation?

### CODE REVIEW

File: dask/array/core.py [hunk 1]
```diff
@@ -1059,14 +1059,17 @@ def elemwise(op, *args, **kwargs):
     arrays = [arg for arg in args if isinstance(arg, Array)]
     other = [(i, arg) for i, arg in enumerate(args) if not isinstance(arg, Array)]
 
-    if not all(a._dtype is not None for a in arrays):
+    if 'dtype' in kwargs:
+        dt = kwargs['dtype']
+    elif not all(a._dtype is not None for a in arrays):
```
Comment (shoyer) [inline dask/array/core.py]: This is the same as `elif any(a._dtype is None for a in arrays)`, I think?

That would avoid the double-negative

File: dask/array/core.py [hunk 2]
```diff
@@ -1106,8 +1109,8 @@ def wrap_elemwise(func):
 fmod = wrap_elemwise(np.fmod)
 frexp = wrap_elemwise(np.frexp)
 hypot = wrap_elemwise(np.hypot)
-isinf = wrap_elemwise(np.isinf)
-isnan = wrap_elemwise(np.isnan)
+isinf = wrap_elemwise(np.isinf, dtype='bool')
+isnan = wrap_elemwise(np.isnan, dtype='bool')
```
Comment (shoyer) [inline dask/array/core.py]: `isnull` and `notnull` are also always `dtype=bool`

**Comment by mrocklin** [comment]
OK, tests now cover all elementwise operations (I think) and dtype checking is tied into all tests.

**Comment by mrocklin** [comment]
Merging this soon if no comment.

**Pull request merged**